client/src/procedures/types.rs

```diff
@@ -373,297 +50,179 @@ pub struct Products<T> {
+    }
 }
 
-impl<T: TargetInfo> PersistSecret for T {
-    fn target(&self) -> Location {
-        self.target_info().write_to.location.clone()
-    }
+/// Output of a [`StrongholdProcedure`][super::StrongholdProcedure].
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct ProcedureOutput(Vec<u8>);
 
-    fn write_secret(mut self, location: Location, hint: RecordHint) -> Self
-    where
-        Self: Sized,
-    {
-        let target = self.target_info_mut();
-        target.write_to = Target { location, hint };
-        target.is_temp = false;
-        self
+impl From<()> for ProcedureOutput {
+    fn from(_: ()) -> Self {
+        ProcedureOutput(Vec::new())
     }
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct Target {
-    pub location: Location,
-    pub hint: RecordHint,
-}
-
-impl Target {
-    pub fn random() -> Self {
-        let location = Location::generic("TEMP".as_bytes(), random::bytestring(32));
-        let hint = RecordHint::new("".to_string()).unwrap();
-        Target { location, hint }
+impl From<Vec<u8>> for ProcedureOutput {
+    fn from(v: Vec<u8>) -> Self {
+        ProcedureOutput(v)
     }
 }
 
-/// Non-secret input for a procedure.
-pub trait InputInfo {
-    type In;
-
-    fn input_info(&self) -> &InputData<Self::In>;
-    fn input_info_mut(&mut self) -> &mut InputData<Self::In>;
+impl From<String> for ProcedureOutput {
+    fn from(s: String) -> Self {
+        s.into_bytes().into()
+    }
 }
 
-/// Specify where non-secret output should be written to.
-pub trait OutputInfo {
-    fn output_info(&self) -> &TempOutput;
-    fn output_info_mut(&mut self) -> &mut TempOutput;
+impl<const N: usize> From<[u8; N]> for ProcedureOutput {
+    fn from(a: [u8; N]) -> Self {
+        a.to_vec().into()
+    }
 }
 
-/// Trait that is auto-implemented for all procedures that have non-secret data output.
-/// Per default, this output is only temporary and will not be in the [`CollectedOutput`]
-/// that is returned to the user.
-/// Temporary output is still cached during the procedure's execution and may be used as input for
-/// subsequent procedures in a chain.
-///
-/// If the output should be returned to the user, an [`OutputKey`] can be set via
-/// [`PersistOutput::store_output`], the same key can then be used to take the output from the
-/// [`CollectedOutput`].
-pub trait PersistOutput {
-    /// [`OutputKey`] with which the procedure's output is associated.
-    /// This key can be used to specify the input of subsequent procedures in a procedure chain.
-    ///
-    /// If [`PersistOutput::store_output`] was not set, this key is random, and the output will not be
-    /// persisted in the [`CollectedOutput`] that is returned to the user.
-    fn output_key(&self) -> OutputKey;
-
-    /// Write the procedure's non-secret output into the [`CollectedOutput`] that is returned to the user.
-    fn store_output(self, key: OutputKey) -> Self;
+impl From<ProcedureOutput> for () {
```

Comment:
hm I guess you have to provide the impl for an empty tuple, since the trait `Procedure` requires `type Output: TryFrom<ProcedureOutput>;` Other option would be to reduce the bound only to `ProcedureOutput`, call `into` would be done anyways and be present for all types, that are allowed as input.

Reply:
> Other option would be to reduce the bound only to `ProcedureOutput`

Do I understand you correctly that you propose that all Procedures should just have the return type `ProcedureOutput`?

The reason why we have the `Procedure::Output` associated type is so that the user can then safely convert it to the actual procedures output type. So if a Procedure has return type `String`, the value is intermediately converted to `ProcedureOutput`, but the user can then later convert it back to `String` and rely on the fact that this won't fail.
This is also done in `Stronghold::exec_proc`, where it enables that the return value of the method is the one of `Procedure::Output`. So that `stronghold.runtime_exec(WriteVault {..}) ` has return value `()`.
https://github.com/iotaledger/stronghold.rs/blob/2fc146afd8862bf5023f9f6e9e9b0828610f7f81/client/src/interface.rs#L294-L301

Reply:
Makes sense, since output types are required to implement a conversion.